[develop] Add process lightning task

EdwardSnyder-NOAA opened the pull request:
## DESCRIPTION OF CHANGES: 
<!-- One or more paragraphs describing the problem, solution, and required changes. -->
This PR adds the process lightning task to the SRW App from the NOAA-GSL RRFS_dev1 branch. The process lightning task is part of the RRFS configuration of the SRW App. 

A wrapper script was added since this task can be ran outside of the workflow. The following are the instructions on how to run this process on Jet:

- Clone repo and build App
- salloc compute node by doing:  salloc -N 1 -n 1 -A hfv3gfs -t 00:30:00 -q batch
- Load envs
  - Module use /path/to/ufs_srweather_app/modulefiles
  - Module load build_jet_intel
    - Note: only need the hpc, hpc-intel, hpc-impi modules so: 
    - module use /mnt/lfs4/HFIP/hfv3gfs/role.epic/hpc-stack/libs/intel-2022.1.2/modulefiles/stack
    - Module load hpc hpc-intel hpc-impi
  - Module load wflow_jet
  - Conda activate regional_workflow
- Create EXPT
  - cd into ush and copy config.community.yaml to config.yaml
  - Update the following vars:
    - MACHINE: jet
    - ACCOUNT: hfv3gfs
    - PREDEF_GRID_NAME: RRFS_CONUS_3km
    - DATE_FIRST_CYCL: '2022072003'
    - DATE_LAST_CYCL: '2022072003'
    - FCST_LEN_HRS: 6
  - Add a rrfs section at the end of the config.yaml file with DO_NLDN_LGHT variable. See example below:
    - rrfs:
      - DO_NLDN_LGHT: true
  - Run generate_FV3LAM_wflow.py script to create expt
- Run EXPT
  - cp run_process_lightning.sh wrapper from ush/wrappers to expt_dir
  - Export var: export EXPTDIR=$PWD
  - Run script: ./run_procs_lightning.sh 2>&1 | tee proc_lght.log


### Type of change
<!-- Please delete options that are not relevant. Add an X to check off a box. -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## TESTS CONDUCTED: 
<!-- Explicitly state what tests were run on these changes, or if any are still pending (for README or other text-only changes, just put "None required"). Make note of the compilers used, the platform/machine, and other relevant details as necessary. For more complicated changes, or those resulting in scientific changes, please be explicit! -->
<!-- Add an X to check off a box. -->

- [ ] hera.intel
- [ ] orion.intel
- [ ] cheyenne.intel
- [ ] cheyenne.gnu
- [ ] gaea.intel
- [x] jet.intel
- [ ] wcoss2.intel
- [ ] NOAA Cloud (indicate which platform)
- [ ] Jenkins
- [ ] fundamental test suite
- [ ] comprehensive tests (specify *which* if a subset was used)

## DEPENDENCIES:
<!-- Add any links to external PRs (e.g. regional_workflow and/or UFS PRs). For example:
- ufs-community/regional_workflow/pull/<pr_number>
- ufs-community/UFS_UTILS/pull/<pr_number>
- ufs-community/ufs-weather-model/pull/<pr_number> -->
None.

## DOCUMENTATION:
<!-- If this PR is contributing new capabilities that need to be documented, please also include updates to the RST files (docs/UsersGuide/source) as supporting material. -->


## ISSUE: 
<!-- If this PR is resolving or referencing one or more issues, in this repository or elsewhere, list them here (Remember, issues must always be created before starting work on a PR branch!). For example, "Fixes issue mentioned in #123" or "Related to bug in https://github.com/ufs-community/other_repository/pull/63" -->

## CHECKLIST
<!-- Add an X to check off a box. -->
- [x] My code follows the style guidelines in the Contributor's Guide
- [x] I have performed a self-review of my own code using the Code Reviewer's Guide
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] My changes need updates to the documentation. I have made corresponding changes to the documentation
- [x] My changes do not require updates to the documentation (explain).
- [x] My changes generate no new warnings
- [ ] New and existing tests pass with my changes
- [ ] Any dependent changes have been merged and published

## LABELS (optional): 
<!-- If you do not have permissions to add labels to your own PR, request that labels be added here. 
Add an X to check off a box. Delete any unnecessary labels. -->
A Code Manager needs to add the following labels to this PR: 
- [ ] Work In Progress
- [ ] bug
- [x] enhancement
- [ ] documentation
- [ ] release
- [ ] high priority
- [ ] run_ci
- [ ] run_we2e_fundamental_tests
- [ ] run_we2e_comprehensive_tests
- [ ] Needs Cheyenne test 
- [ ] Needs Jet test 
- [ ] Needs Hera test 
- [ ] Needs Orion test 
- [ ] help wanted

## CONTRIBUTORS (optional): 
<!-- If others have contributed to this work aside from the PR author, list them here -->

Commit:
Added rrfs_utils to the cmake settings

Commit:
Added jobs and exregional lightning scripts

Commit:
Added lightning parm

christinaholtNOAA commented on jobs/JREGIONAL_PROCESS_LIGHTNING:
```diff
@@ -0,0 +1,96 @@
+#-----------------------------------------------------------------------
+#
+# This J-JOB script runs the NetCDF ligthning observation preprocess 
+#      for the FV3-LAM model
```
ligthning -> lightning

EdwardSnyder-NOAA replied:
updated.

christinaholtNOAA commented on scripts/exregional_process_lightning.sh:
```diff
@@ -0,0 +1,207 @@
+In directory:     \"${scrfunc_dir}\"
+
+This is the ex-script for the task that runs lightning preprocess
+with FV3 for the specified cycle.
```
preprocess -> preprocessing

EdwardSnyder-NOAA replied:
updated.

christinaholtNOAA commented on scripts/exregional_process_lightning.sh:
```diff
@@ -0,0 +1,207 @@
+#
+#-----------------------------------------------------------------------
+#
+START_DATE=$(echo "${CDATE}" | sed 's/\([[:digit:]]\{2\}\)$/ \1/')
```
I don't think we have `CDATE` anymore. The date is now passed with NCO-compliant `PDY` and `cyc` variables. For the timestamp variables below, I think that translates to:

```
YYYYMMDDHH=${PDY}${cyc}
YYYYMMDD=${PDY}
HH=${cyc}
```

EdwardSnyder-NOAA replied:
I'm unsure what the NCO-complaint variables are. But in the develop branch, `CDATE` is used in a number of `ufs-srweather-app/scripts/*` and CDATE is defined in the `job_preamble.sh` script as `export CDATE=${PDY}${cyc}`.

MatthewPyle-NOAA replied:
I would consider `CDATE` to be non-standard, but tolerated in operations (is pretty widely used in the GFS).

EdwardSnyder-NOAA replied:
Ah okay. That's good to know. I switched the code to be: `START_DATE=$(echo "${PDY} ${cyc}")`

christinaholtNOAA commented on scripts/exregional_process_lightning.sh:
```diff
@@ -0,0 +1,207 @@
+#-----------------------------------------------------------------------
+run_lightning=false
+filenum=0
+LIGHTNING_FILE=${LIGHTNING_ROOT}/vaisala/netcdf
```
Where is `LIGHTNING_ROOT` set? I assume in the `var_defns.yaml` file since its capitalized?

EdwardSnyder-NOAA replied:
`LIGHTNING_ROOT` is set in the `var_defns.yaml` script.

christinaholtNOAA commented on scripts/exregional_process_lightning.sh:
```diff
@@ -0,0 +1,207 @@
+YYYYMMDD=${YYYYMMDDHH:0:8}
+
+YYJJJHH=$(date +"%y%j%H" -d "${START_DATE}")
+PREYYJJJHH=$(date +"%y%j%H" -d "${START_DATE} 1 hours ago")
```
Would you mind changing this variable to `EARLY_YYJJJHH`? I can't not read "PREY"

I was about to suggest "PREV" but realized that terminology is tied pretty heavily with hourly cycling. We'd still want these observations before the top of the hour, even if we were cycling every 6 hours, I think.

If some other prefix makes more sense, please feel free to choose that instead.

EdwardSnyder-NOAA replied:
Updated to `EARLY_YYJJJHH`.

christinaholtNOAA commented on scripts/exregional_process_lightning.sh:
```diff
@@ -0,0 +1,207 @@
+  else
+   echo " ${filename} does not exist"
+  fi
+done
```
This is all essentially the same logic, just stepping from 25 minutes before the hour to 5 mins after....I'm curious if it might lend itself better to a single over time?

```
for incr in $(seq -25 5 5) ; do 
  filedate=$(date +"%y%j%H%M" -d "${START_DATE} ${incr} minutes ")
  filename=${LIGHTNING_FILE}/${filedate}0005r
  ...
done

```


With this approach, that might mean a significant cleanup of the date variables needed earlier in this script.

EdwardSnyder-NOAA replied:
Switched to this block of code and removed a number of date variables we won't need.

christinaholtNOAA commented on scripts/exregional_process_lightning.sh:
```diff
@@ -0,0 +1,207 @@
+if [[ "$run_lightning" == true ]]; then
+    PREP_STEP
+    eval $RUN_CMD_UTILS --account=${ACCOUNT} --nodes=1-1 --tasks-per-node=1 --time 00:30:00 ${exec_fp} ${REDIRECT_OUT_ERR} || print_err_msg "\  
+        Call to executable to run lightning (nc) process returned with nonzero exit code."
```
This seems like more args than we typically pass a `$RUN_CMD variable`.

EdwardSnyder-NOAA replied:
I agree and I removed the args we don't need.

EdwardSnyder-NOAA replied:
I now realized those extra args were there for me to run this task without logging onto a compute node.

christinaholtNOAA reviewed (commented):
@EdwardSnyder-NOAA This is shaping up nicely. I have left a few comments/suggestions below.

I'm curious if you made changes to the default_configs.yaml file. Perhaps it's missing here? Also, if you tested the script standalone, would you mind including the necessary environment variables you passed the j-job in the j-job header?

Commit:
did some cleanup

Commit:
added lightning wrapper

Commit:
added rrfs lightning vars

Commit:
add fix gsi path

EdwardSnyder-NOAA committed:
Merge branch 'ufs-community:develop' into feature/rrfs-lightning

Commit:
Merge branch 'feature/rrfs-lightning' of https://github.com/EdwardSnyder-NOAA/ufs-srweather-app into feature/rrfs-lightning

Commit:
removed fixgsi dir

EdwardSnyder-NOAA commented:
> @EdwardSnyder-NOAA This is shaping up nicely. I have left a few comments/suggestions below.
> 
> I'm curious if you made changes to the default_configs.yaml file. Perhaps it's missing here? Also, if you tested the script standalone, would you mind including the necessary environment variables you passed the j-job in the j-job header?

@christinaholtNOAA I added the default_config.yaml file and the wrapper script I used to run this process. I also included instructions on how to run this task with it. In addition, I addressed your comments from earlier. Let me know what you think. I just noticed I still have a number of date variables commented out in the exregional script, which can be removed.

Commit:
updated path to predefined grid for fv3_grid_spec

Commit:
added lightning root and fixgsi vars

Commit:
updated DOMAIN_PREGEN_BASEDIR to point to EPIC location

christinaholtNOAA commented on scripts/exregional_process_lightning.sh:
```diff
@@ -0,0 +1,204 @@
+
+#YYJJJHH=$(date +"%y%j%H" -d "${START_DATE}")
+#EARLY_YYJJJHH=$(date +"%y%j%H" -d "${START_DATE} 1 hours ago")
+
```
Please remove commented code.

christinaholtNOAA commented on ush/machine/hera.yaml:
```diff
@@ -29,13 +29,14 @@ platform:
   FIXshp: /scratch2/BMC/det/UFS_SRW_App/develop/NaturalEarth
   EXTRN_MDL_DATA_STORES: hpss aws nomads
-
+  LIGHTNING_ROOT: /scratch2/BMC/zrtrr/rli/data/lightning
```
We don't want to support this location for the whole community.

EdwardSnyder-NOAA replied:
updated path to /scratch1/NCEPDEV/nems/role.epic/UFS_SRW_data/develop/rrfs_retro_data/lightning/vaisala/netcdf

christinaholtNOAA commented on ush/machine/jet.yaml:
```diff
@@ -26,12 +26,14 @@ platform:
   FIXsfc: /mnt/lfs4/BMC/wrfruc/UFS_SRW_App/develop/fix/fix_sfc_climo
   FIXshp: /mnt/lfs4/BMC/wrfruc/UFS_SRW_App/develop/NaturalEarth
   EXTRN_MDL_DATA_STORES: hpss aws nomads
+  LIGHTNING_ROOT: /mnt/lfs4/BMC/wrfruc/Ruifang.Li/data/lightning
```
This one is probably also not good to support.

EdwardSnyder-NOAA replied:
@christinaholtNOAA - Are we wanting to store this lightning data under our EPIC or SRW controlled location?

EdwardSnyder-NOAA replied:
updated path to /mnt/lfs4/HFIP/hfv3gfs/role.epic/UFS_SRW_data/develop/rrfs_retro_data/lightning/vaisala/netcdf

christinaholtNOAA reviewed (changes requested):
I think it would be good to stage some lightning data for testing with the other test data, but we'd likely want to preserve the lightning data root variable for any supported near-real time data streams.

Commit:
removed unneeded date paths

EdwardSnyder-NOAA committed:
Merge branch 'develop' into feature/rrfs-lightning

Commit:
updated lightning root var to epic location

Commit:
removed unneeded lightning var

Commit:
removed lightning root

EdwardSnyder-NOAA commented:
> I think it would be good to stage some lightning data for testing with the other test data, but we'd likely want to preserve the lightning data root variable for any supported near-real time data streams.

I added some lightning data to Hera and Jet to test lightning functionality and it worked. This PR is ready for review.

EdwardSnyder-NOAA committed:
Merge branch 'develop' into feature/rrfs-lightning

christinaholtNOAA reviewed (approved):
LGTM.

christinaholtNOAA commented:
@MichaelLueken We'd also like to focus on getting this one merged for RRFS, without waiting on any others. Thanks!

MichaelLueken commented:
@EdwardSnyder-NOAA I'm working on reviewing this PR now and see that there is a conflict in `ush/config_defaults.yaml`.  Please merge the latest develop into your `feature/rrfs-lightning` branch to address the conflict at your earliest convenience.  Thanks!

EdwardSnyder-NOAA committed:
Merge branch 'develop' into feature/rrfs-lightning

EdwardSnyder-NOAA commented on ush/config_defaults.yaml:
```diff
@@ -1031,6 +1046,21 @@ nco:
+  #-----------------------------------------------------------------------
+  #
+  DO_NLDN_LGHT: false
+
```
@christinaholtNOAA in #647 you mentioned that these do switches at this level and quantity shouldn't be how the workflow is configured, and that we should define the tasks based on its definition. If that's the case, do I need to include the `DO_NLDN_LGHT` variable?

christinaholtNOAA replied:
I'm fine with it staying. I think we can assess whether it's needed for the workflow when it's integrated. One of these here or there can be handled. It's hard to balance/manage a bunch of them, though.

EdwardSnyder-NOAA commented:
@MichaelLueken thanks for catching that! Conflict should be resolved now. While resolving the conflict, I noticed some logic I may not need, so I tagged Christina for assistance.

MichaelLueken reviewed (approved):
@EdwardSnyder-NOAA These changes look good to me and I was able to successfully run the wrapper script on Jet.  I will now approve these changes, but will hold off on running the Jenkins tests on this work until @christinaholtNOAA has confirmed whether to include `DO_NLDN_LGHT` in `ush/config_defaults.yaml`.

MichaelLueken commented:
@EdwardSnyder-NOAA Just to check, when this work is merged to develop, then issue #487 will be complete, correct?  Thanks!

christinaholtNOAA commented:
@MichaelLueken 👍 from me.

MichaelLueken commented:
The Jenkins tests completed.  The tests on Hera Intel failed and `grid_RRFS_CONUS_25km_ics_NAM_lbcs_NAM_suite_GFS_v16` failed on Cheyenne Intel.  The failed test was manually ran on Cheyenne and successfully passed.  The Hera Intel tests were ran manually and passed (with the exception of `grid_RRFS_CONUS_25km_ics_FV3GFS_lbcs_RAP_suite_HRRR` due to issue #688).  Proceeding with merging this work to develop.
